[cxx-interop] noexcept specifier before function attributes

drodriguez opened the pull request:
In #74516 the `SWIFT_NOEXCEPT` specifier is added to the imported Swift functions in C++ mode, but it is added after the function attributes. It seems that the tests only do `-fsyntax-only`, which seems not to catch an error like "expected function body after function declarator" when the header is used without that flag.

Flip the attributes and the specifier around in the printer, and flip them in all the tests.

drodriguez committed:
[cxx-interop] noexcept specifier before function attributes

In #74516 the `SWIFT_NOEXCEPT` specifier is added to the imported Swift
functions in C++ mode, but it is added after the function attributes. It seems
that the tests only do `-fsyntax-only`, which seems not to catch an error like
"expected function body after function declarator" when the header is
used without that flag.

Flip the attributes and the specifier around in the printer, and flip
them in all the tests.

drodriguez commented:
@swift-ci please smoke test

Xazax-hun reviewed (approved):
LGTM! Thanks for catching this!

Xazax-hun commented:
I think this also needs to be cherry-picked to `release/6.0`.

drodriguez commented:
I checked the other usages of `SWIFT_NOEXCEPT` and in every case it seems that they are added inmediately after closing the parenthesis of the function arguments, so they should be good.

drodriguez committed:
[cxx-interop] Add tests for PrintAsObjC with C++ context

The tests were using `%check-in-clang`, but it only checks importing as
an objective-c-header. Add a parallel `%check-in-clang-cxx` to test also
in C++. It uses C++17 because of some details in the imported headers and
disables a warning about variadic macros that was promoted to an error
and was blocking passing the tests. The clang-importer-sdk gets the
minimal set of files to compile the two modified tests as C++. The files
are mostly empty, except `cstddef` that imports the equivalent C header.
Some modifications were needed in `ctypes.h` because the header was
using features only available in C and not C++.

drodriguez commented:
@swift-ci please test

drodriguez commented:
Added a second commit with the test modified to also do the check with a C++ context. Before it was only done as a C header, which simply did not have anything where `noexcept` should had gone, passing the test.

drodriguez commented:
@swift-ci please test Linux platform

drodriguez commented:
Windows CI is not having a great day https://ci-external.swift.org/job/swift-PR-windows/buildTimeTrend

I will ignore it and hope it lets me merge.

drodriguez commented:
@swift-ci please test Windows platform
